### ISSUE TITLE
Question about the triplet loss

### ISSUE THREAD

**Issue by gpantaz** [UKPLab/sentence-transformers#3257]
Hello!

I would like to use the triplet loss with hard negatives but I am not sure whether I should normalize the embeddings before computing the pairwise distances. The default implementation uses the euclidean distance between unnormalized embeddings but I do not understand why.

https://github.com/UKPLab/sentence-transformers/blob/309a9accb784e0931636ad97be7806e37c57dfb4/sentence_transformers/losses/BatchHardTripletLoss.py#L58-L64

Any thoughts on why this is happening are much appreciated, thanks! :)

**Comment by tomaarsen** [comment]
Hello!

The Euclidean (I see now that there's a typo, unfortunate), distance is the default distance used in the original [paper](https://arxiv.org/pdf/1703.07737), as well as other [implementations](https://github.com/NegatioN/OnlineMiningTripletLoss/blob/4c22eac01ee87d4e5a79f3a91df01d268cf249ec/online_triplet_loss/losses.py#L112-L113). 

The author of the paper suggests to not use a normalization layer when training with this loss:
> Note that when forcing the embedding’s norm to one,
using the squared Euclidean distance corresponds to using
the cosine-similarity, up to a constant factor of two. We did
not use a normalizing layer in any of our final experiments.
For one, it does not dramatically regularize the network by
reducing the available embedding space: the space spanned
by all D-dimensional vector of fixed norm is still a D − 1-
dimensional volume. Worse, an output-normalization layer
can actually hide problems in the training, such as slowly
collapsing or exploding embeddings.

I hope this helps explain it somewhat.

- Tom Aarsen

**Comment by tomaarsen** [comment]
> Hi Tom, thank you for your fast response, I did notice indeed that the authors have included this paragraph but I was wondering if there are any quantitive results that show this. In my case I did notice a substantial improvement when normalizing the embeddings before computing the distances and was curious to why this happening.

I haven't ran the `Batch...TripletLoss` classes a lot - in my experience it's a bit hard to get data for them (as you can see in the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html), they require text & class data).

Instead, I generally have data with actual triplets: `anchor, positive, negative`, at which point I prefer the [TripletLoss](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#tripletloss) or the [MultipleNegativesRankingLoss](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#multiplenegativesrankingloss) (or a variant). Especially the latter is very strong, as it *also* gets in-batch negatives in addition to the hard negatives. 
If the explanation of the MNRL loss is a bit confusing on the website, then perhaps this one is clearer:
https://github.com/tomaarsen/sentence-transformers/blob/2ec81c5fe0217d1e1c4484b0629a002d716e2b01/sentence_transformers/losses/MultipleNegativesRankingLoss.py#L16-L29

> Also, happy to fix the typo in a PR :)

Sadly, it's tricky to fix, as fixing it outright will break older training scripts. That would break backwards compatibility, which I want to avoid as much as possible.

- Tom Aarsen

**Comment by gpantaz** [comment]
Thanks for the info will have a look at the other losses as well :)
